File: .vendor/flamewm-c-reference/src/flamewm/platform/host.cc
#include "flamewm/platform/host.h"

#include <cstddef>
#include <string>

namespace flamewm {
namespace platform {

struct PlatformHost::Impl {
    settings::SettingsService* settings;
    reactor::ReactorService* reactor;
    windows::WindowService* windows;
    workspaces::WorkspaceService* workspaces;
    displays::DisplayService* displays;
    shortcuts::ShortcutService* shortcuts;
    panels::PanelService* panels;
    applications::ApplicationService* applications;
    scale::ScaleService* scale;
    session::SessionService* session;
    background::BackgroundService* background;
    system::SystemService* system;
    snap::SnapService* snap;
    chrome::ChromePolicy* chrome;

    Impl()
        : settings(nullptr)
        , reactor(nullptr)
        , windows(nullptr)
        , workspaces(nullptr)
        , displays(nullptr)
        , shortcuts(nullptr)
        , panels(nullptr)
        , applications(nullptr)
        , scale(nullptr)
        , session(nullptr)
        , background(nullptr)
        , system(nullptr)
        , snap(nullptr)
        , chrome(nullptr)
    {}
};

PlatformHost::PlatformHost(api::EnginePorts ports)
    : ports_(ports)
    , config_()
    , generation_(0)
    , started_(false)
    , impl_(nullptr)
{}

PlatformHost::PlatformHost(api::EnginePorts ports, const Config& config)
    : ports_(ports)
    , config_(config)
    , generation_(0)
    , started_(false)
    , impl_(nullptr)
{}

PlatformHost::~PlatformHost() {
    if (started_) {
        stop();
    } else if (impl_) {
        delete impl_;
        impl_ = nullptr;
    }
}

bool PlatformHost::validatePorts() const {
    return ports_.window      != nullptr
        && ports_.workspace   != nullptr
        && ports_.display     != nullptr
        && ports_.shortcut    != nullptr
        && ports_.workArea    != nullptr
        && ports_.mainLoop    != nullptr
        && ports_.input       != nullptr
        && ports_.application != nullptr
        && ports_.session     != nullptr
        && ports_.background  != nullptr
        && ports_.tray        != nullptr;
}

bool PlatformHost::start() {
    if (started_) {
        return false;
    }
    if (!validatePorts()) {
        return false;
    }

    Impl* impl = new Impl();

    impl->settings = new settings::SettingsService(config_.settingsPath);
    impl->reactor  = new reactor::ReactorService(ports_.mainLoop);
    impl->scale    = new scale::ScaleService();
    impl->chrome   = new chrome::ChromePolicy();
    impl->windows  = new windows::WindowService(ports_.window);
    impl->workspaces = new workspaces::WorkspaceService(ports_.workspace);
    impl->displays = new displays::DisplayService(ports_.display, ports_.mainLoop);
    impl->shortcuts = new shortcuts::ShortcutService(ports_.shortcut);
    impl->panels   = new panels::PanelService(ports_.workArea, ports_.tray);
    impl->applications = new applications::ApplicationService(ports_.application);
    impl->session  = new session::SessionService(ports_.session);
    if (!config_.settingsPath.empty()) {
        impl->background = new background::BackgroundService(
            ports_.background, config_.enginePreferencesPath, config_.settingsPath);
    } else {
        impl->background = new background::BackgroundService(
            ports_.background, config_.enginePreferencesPath);
    }
    impl->system   = new system::SystemService(impl->reactor);
    impl->snap     = new snap::SnapService(ports_.window, ports_.mainLoop);

    impl_ = impl;
    started_ = true;
    return true;
}

void PlatformHost::stop() {
    if (!started_) {
        if (impl_) {
            delete impl_;
            impl_ = nullptr;
        }
        return;
    }

    ++generation_;

    if (impl_) {
        if (impl_->reactor) impl_->reactor->stopAll();
        if (impl_->background) { /* no clearSubscriptions API */ }
        if (impl_->panels) impl_->panels->clearSubscriptions();
        if (impl_->scale) impl_->scale->clearSubscriptions();
        if (impl_->chrome) impl_->chrome->clearSubscriptions();
    }

    if (impl_) {
        delete impl_->snap;         impl_->snap = nullptr;
        delete impl_->system;       impl_->system = nullptr;
        delete impl_->background;   impl_->background = nullptr;
        delete impl_->session;      impl_->session = nullptr;
        delete impl_->applications; impl_->applications = nullptr;
        delete impl_->panels;       impl_->panels = nullptr;
        delete impl_->shortcuts;    impl_->shortcuts = nullptr;
        delete impl_->displays;     impl_->displays = nullptr;
        delete impl_->workspaces;   impl_->workspaces = nullptr;
        delete impl_->windows;      impl_->windows = nullptr;
        delete impl_->chrome;       impl_->chrome = nullptr;
        delete impl_->scale;        impl_->scale = nullptr;
        delete impl_->reactor;      impl_->reactor = nullptr;
        delete impl_->settings;     impl_->settings = nullptr;

        delete impl_;
        impl_ = nullptr;
    }

    started_ = false;
}

windows::WindowService* PlatformHost::windows() { return impl_ ? impl_->windows : nullptr; }
workspaces::WorkspaceService* PlatformHost::workspaces() { return impl_ ? impl_->workspaces : nullptr; }
settings::SettingsService* PlatformHost::settings() { return impl_ ? impl_->settings : nullptr; }
shortcuts::ShortcutService* PlatformHost::shortcuts() { return impl_ ? impl_->shortcuts : nullptr; }
displays::DisplayService* PlatformHost::displays() { return impl_ ? impl_->displays : nullptr; }
panels::PanelService* PlatformHost::panels() { return impl_ ? impl_->panels : nullptr; }
applications::ApplicationService* PlatformHost::applications() { return impl_ ? impl_->applications : nullptr; }
scale::ScaleService* PlatformHost::scale() { return impl_ ? impl_->scale : nullptr; }
session::SessionService* PlatformHost::session() { return impl_ ? impl_->session : nullptr; }
background::BackgroundService* PlatformHost::background() { return impl_ ? impl_->background : nullptr; }
reactor::ReactorService* PlatformHost::reactor() { return impl_ ? impl_->reactor : nullptr; }
system::SystemService* PlatformHost::system() { return impl_ ? impl_->system : nullptr; }
snap::SnapService* PlatformHost::snap() { return impl_ ? impl_->snap : nullptr; }
chrome::ChromePolicy* PlatformHost::chrome() { return impl_ ? impl_->chrome : nullptr; }

const windows::WindowService* PlatformHost::windows() const { return impl_ ? impl_->windows : nullptr; }
const workspaces::WorkspaceService* PlatformHost::workspaces() const { return impl_ ? impl_->workspaces : nullptr; }
const settings::SettingsService* PlatformHost::settings() const { return impl_ ? impl_->settings : nullptr; }
const shortcuts::ShortcutService* PlatformHost::shortcuts() const { return impl_ ? impl_->shortcuts : nullptr; }
const displays::DisplayService* PlatformHost::displays() const { return impl_ ? impl_->displays : nullptr; }
const panels::PanelService* PlatformHost::panels() const { return impl_ ? impl_->panels : nullptr; }
const applications::ApplicationService* PlatformHost::applications() const { return impl_ ? impl_->applications : nullptr; }
const scale::ScaleService* PlatformHost::scale() const { return impl_ ? impl_->scale : nullptr; }
const session::SessionService* PlatformHost::session() const { return impl_ ? impl_->session : nullptr; }
const background::BackgroundService* PlatformHost::background() const { return impl_ ? impl_->background : nullptr; }
const reactor::ReactorService* PlatformHost::reactor() const { return impl_ ? impl_->reactor : nullptr; }
const system::SystemService* PlatformHost::system() const { return impl_ ? impl_->system : nullptr; }
const snap::SnapService* PlatformHost::snap() const { return impl_ ? impl_->snap : nullptr; }
const chrome::ChromePolicy* PlatformHost::chrome() const { return impl_ ? impl_->chrome : nullptr; }

} // namespace platform
} // namespace flamewm

File: crates/flamewm-platform/src/host.rs
use std::path::PathBuf;

use flamewm_api::applications::{ApplicationLaunchOptions, DesktopApplication};
use flamewm_api::capabilities::Capabilities;
use flamewm_api::display::DisplaySnapshot;
use flamewm_api::panels::PanelsSnapshot;
use flamewm_api::ports::EnginePorts;
use flamewm_api::session::{SessionAction, SessionCapabilities};
use flamewm_api::settings::{SettingsSnapshot, SettingsTransaction};
use flamewm_api::shortcuts::{KeyBinding, ShortcutSnapshot};
use flamewm_api::system::{SystemAction, SystemSnapshot};
use flamewm_api::wm_features::{
    FeatureAction, FullscreenMonitorSpan, InteractiveMoveResize, RestackMode, WindowFeature,
    WindowFeatureSnapshot,
};
use flamewm_api::workspace::WorkspaceSnapshot;
use flamewm_api::{
    DesktopAppId, FlameResult, ModeId, OutputId, PanelEdge, TaskEntryId, TransactionId, WindowRef,
};

use crate::services::applications::ApplicationService;
use crate::services::background::BackgroundService;
use crate::services::display::DisplayService;
use crate::services::panels::PanelService;
use crate::services::session::SessionService;
use crate::services::settings::SettingsService;
use crate::services::shortcuts::ShortcutService;
use crate::services::windows::WindowService;
use crate::services::workspaces::{NavigationDirection, WorkspaceService};
use crate::system::{SystemActionHandler, SystemService};

/// Process composition root for FlameWM product state.
///
/// The engine adapter is owned once. Product services never retain pointers/references into it;
/// every command borrows the required port for exactly the duration of that command. This avoids
/// the native shutdown/UAF class where views outlived raw service/engine pointers.
pub struct PlatformHost<E: EnginePorts> {
    engine: E,
    started: bool,
    generation: u64,
    windows: WindowService,
    workspaces: WorkspaceService,
    displays: DisplayService,
    panels: PanelService,
    settings: SettingsService,
    shortcuts: ShortcutService,
    applications: ApplicationService,
    session: SessionService,
    background: BackgroundService,
    system: SystemService,
}

impl<E: EnginePorts> PlatformHost<E> {
    #[must_use]
    pub fn new(engine: E, settings_path: PathBuf) -> Self {
        Self {
            engine,
            started: false,
            generation: 0,
            windows: WindowService::default(),
            workspaces: WorkspaceService,
            displays: DisplayService::default(),
            panels: PanelService::default(),
            settings: SettingsService::new(settings_path),
            shortcuts: ShortcutService::default(),
            applications: ApplicationService::default(),
            session: SessionService,
            background: BackgroundService::default(),
            system: SystemService::default(),
        }
    }

    pub fn start(&mut self) -> FlameResult<()> {
        if self.started {
            return Ok(());
        }
        self.windows.refresh(&self.engine)?;
        let displays = self.displays.refresh(&mut self.engine)?.clone();
        self.panels.sync_outputs(&displays);
        self.started = true;
        self.generation = self.generation.saturating_add(1).max(1);
        Ok(())
    }

    pub fn stop(&mut self) {
        if !self.started {
            return;
        }
        self.panels.close_start();
        self.started = false;
        self.generation = self.generation.saturating_add(1).max(1);
    }

    #[must_use]
    pub const fn is_started(&self) -> bool {
        self.started
    }

    #[must_use]
    pub const fn generation(&self) -> u64 {
        self.generation
    }

    #[must_use]
    pub fn capabilities(&self) -> Capabilities {
        Capabilities::current()
    }

    pub fn refresh_windows(&mut self) -> FlameResult<bool> {
        let changed = self.windows.refresh(&self.engine)?;
        if changed {
            let _ = self.panels.reconcile_windows(self.windows.cached());
        }
        Ok(changed)
    }

    pub fn activate_window(&mut self, window: WindowRef) -> FlameResult<()> {
        self.windows.activate(&mut self.engine, window)
    }

    pub fn minimize_window(&mut self, window: WindowRef) -> FlameResult<()> {
        self.windows.minimize(&mut self.engine, window)
    }

    pub fn restore_window(&mut self, window: WindowRef) -> FlameResult<()> {
        self.windows.restore(&mut self.engine, window)
    }

    pub fn close_window(&mut self, window: WindowRef) -> FlameResult<()> {
        self.windows.close(&mut self.engine, window)
    }

    pub fn workspace_snapshot(&self) -> FlameResult<WorkspaceSnapshot> {
        self.workspaces.snapshot(&self.engine)
    }

    pub fn activate_workspace(&mut self, index: usize, expected_revision: u64) -> FlameResult<()> {
        self.workspaces
            .activate(&mut self.engine, index, expected_revision)
    }

    pub fn insert_workspace_after(
        &mut self,
        index: usize,
        expected_revision: u64,
    ) -> FlameResult<()> {
        self.workspaces
            .insert_after(&mut self.engine, index, expected_revision)
    }

    pub fn remove_workspace(&mut self, index: usize, expected_revision: u64) -> FlameResult<()> {
        self.workspaces
            .remove(&mut self.engine, index, expected_revision)
    }

    pub fn navigate_workspace(
        &mut self,
        direction: NavigationDirection,
        expected_revision: u64,
    ) -> FlameResult<()> {
        self.workspaces
            .navigate(&mut self.engine, direction, expected_revision)
    }

    pub fn move_window_to_workspace(
        &mut self,
        window: WindowRef,
        target: usize,
    ) -> FlameResult<()> {
        self.workspaces
            .move_window(&mut self.engine, window, target)
    }

    pub fn window_feature_snapshot(&self, window: WindowRef) -> FlameResult<WindowFeatureSnapshot> {
        self.windows.feature_snapshot(&self.engine, window)
    }

    pub fn apply_window_feature(
        &mut self,
        window: WindowRef,
        feature: WindowFeature,
        action: FeatureAction,
    ) -> FlameResult<()> {
        self.windows
            .apply_feature(&mut self.engine, window, feature, action)
    }

    pub fn showing_desktop(&self) -> FlameResult<bool> {
        self.windows.showing_desktop(&self.engine)
    }

    pub fn set_showing_desktop(&mut self, show: bool) -> FlameResult<()> {
        self.windows.set_showing_desktop(&mut self.engine, show)?;
        let _ = self.refresh_windows()?;
        Ok(())
    }

    pub fn restack_window(
        &mut self,
        window: WindowRef,
        mode: RestackMode,
        sibling: Option<WindowRef>,
    ) -> FlameResult<()> {
        self.windows
            .restack(&mut self.engine, window, mode, sibling)
    }

    pub fn set_fullscreen_monitors(
        &mut self,
        window: WindowRef,
        span: FullscreenMonitorSpan,
    ) -> FlameResult<()> {
        self.windows
            .set_fullscreen_monitors(&mut self.engine, window, span)
    }

    pub fn begin_interactive_move_resize(
        &mut self,
        request: InteractiveMoveResize,
    ) -> FlameResult<()> {
        self.windows
            .begin_interactive_move_resize(&mut self.engine, request)
    }

    pub fn cancel_interactive_move_resize(&mut self, window: WindowRef) -> FlameResult<()> {
        self.windows
            .cancel_interactive_move_resize(&mut self.engine, window)
    }

    pub fn display_snapshot(&mut self) -> FlameResult<DisplaySnapshot> {
        Ok(self.displays.refresh(&mut self.engine)?.clone())
    }

    /// Reconcile per-output panel surfaces/reservations after a display topology refresh.
    pub fn resync_panels_for_displays(&mut self, displays: &DisplaySnapshot) -> FlameResult<bool> {
        Ok(self.panels.sync_outputs(displays))
    }

    pub fn begin_display_mode(
        &mut self,
        output: &OutputId,
        mode: ModeId,
        expected_generation: u64,
        now_ms: u64,
    ) -> FlameResult<TransactionId> {
        self.displays
            .begin_mode(&mut self.engine, output, mode, expected_generation, now_ms)
    }

    pub fn keep_display_mode(&mut self, transaction: TransactionId) -> FlameResult<()> {
        self.displays.keep(&mut self.engine, transaction)
    }

    pub fn revert_display_mode(&mut self, transaction: TransactionId) -> FlameResult<()> {
        self.displays.revert(&mut self.engine, transaction)
    }

    pub fn tick_display_transactions(&mut self, now_ms: u64) -> FlameResult<bool> {
        self.displays.tick(&mut self.engine, now_ms)
    }

    pub fn set_shell_scale(
        &mut self,
        output: &OutputId,
        percent: u16,
        expected_generation: u64,
    ) -> FlameResult<bool> {
        self.displays
            .set_shell_scale(output, percent, expected_generation)
    }

    #[must_use]
    pub fn settings_snapshot(&self) -> SettingsSnapshot {
        self.settings.snapshot()
    }

    pub fn apply_settings(
        &mut self,
        transaction: &SettingsTransaction,
    ) -> FlameResult<SettingsSnapshot> {
        self.settings.apply(&mut self.engine, transaction)
    }

    #[must_use]
    pub fn shortcut_snapshot(&self) -> ShortcutSnapshot {
        self.shortcuts.snapshot()
    }

    pub fn set_shortcut(
        &mut self,
        action: &str,
        binding: &str,
        expected_revision: u64,
    ) -> FlameResult<()> {
        self.shortcuts
            .set_binding(&mut self.engine, action, binding, expected_revision)
    }

    pub fn clear_shortcut(&mut self, action: &str, expected_revision: u64) -> FlameResult<()> {
        self.shortcuts
            .clear_binding(&mut self.engine, action, expected_revision)
    }

    pub fn reset_shortcut(&mut self, action: &str, expected_revision: u64) -> FlameResult<()> {
        self.shortcuts
            .reset_binding(&mut self.engine, action, expected_revision)
    }

    pub fn apply_shortcuts(
        &mut self,
        bindings: &std::collections::BTreeMap<String, KeyBinding>,
        expected_revision: u64,
    ) -> FlameResult<()> {
        self.shortcuts
            .apply(&mut self.engine, bindings, expected_revision)
    }

    #[must_use]
    pub fn panels_snapshot(&self) -> PanelsSnapshot {
        self.panels.snapshot()
    }

    pub fn set_panel_edge(
        &mut self,
        output: &OutputId,
        edge: PanelEdge,
        expected_revision: u64,
    ) -> FlameResult<()> {
        self.panels.set_edge(output, edge, expected_revision)
    }

    pub fn set_panel_size(
        &mut self,
        output: &OutputId,
        logical_size: u16,
        expected_revision: u64,
    ) -> FlameResult<()> {
        self.panels
            .set_size(output, logical_size, expected_revision)
    }

    pub fn pin_app(&mut self, app: DesktopAppId, expected_revision: u64) -> FlameResult<bool> {
        self.panels.pin_app(app, expected_revision)
    }

    pub fn unpin_app(&mut self, app: &DesktopAppId, expected_revision: u64) -> FlameResult<bool> {
        self.panels.unpin_app(app, expected_revision)
    }

    pub fn reorder_task(
        &mut self,
        from: usize,
        to: usize,
        expected_revision: u64,
    ) -> FlameResult<bool> {
        self.panels.reorder(from, to, expected_revision)
    }

    pub fn reorder_task_entry(
        &mut self,
        entry_id: &TaskEntryId,
        to: usize,
        expected_revision: u64,
    ) -> FlameResult<bool> {
        self.panels.reorder_entry(entry_id, to, expected_revision)
    }

    pub fn toggle_start(&mut self) -> bool {
        self.panels.toggle_start()
    }

    pub fn replace_applications(&mut self, applications: Vec<DesktopApplication>) -> bool {
        self.applications.replace_catalog(applications)
    }

    #[must_use]
    pub fn search_applications(&self, query: &str) -> Vec<DesktopApplication> {
        self.applications.search(query)
    }

    pub fn launch_application(
        &mut self,
        app: &DesktopAppId,
        options: &ApplicationLaunchOptions,
    ) -> FlameResult<()> {
        self.applications.launch(&mut self.engine, app, options)
    }

    #[must_use]
    pub fn session_capabilities(&self) -> SessionCapabilities {
        self.session.capabilities(&self.engine)
    }

    pub fn session_action(&mut self, action: SessionAction) -> FlameResult<()> {
        self.session.perform(&mut self.engine, action)
    }

    #[must_use]
    pub fn system_snapshot(&self) -> SystemSnapshot {
        self.system.snapshot().clone()
    }

    pub fn set_system_action_handler(&mut self, handler: SystemActionHandler) {
        self.system.set_action_handler(handler);
    }

    pub fn system_action(
        &mut self,
        action: SystemAction,
        expected_revision: u64,
    ) -> FlameResult<()> {
        self.system.perform_action(action, expected_revision)
    }

    #[must_use]
    pub fn system(&self) -> &SystemService {
        &self.system
    }

    pub fn system_mut(&mut self) -> &mut SystemService {
        &mut self.system
    }

    #[must_use]
    pub fn background(&self) -> &BackgroundService {
        &self.background
    }

    #[must_use]
    pub fn engine(&self) -> &E {
        &self.engine
    }

    pub fn engine_mut(&mut self) -> &mut E {
        &mut self.engine
    }
}

impl<E: EnginePorts> Drop for PlatformHost<E> {
    fn drop(&mut self) {
        self.stop();
    }
}
